hypothesis/lms
# Support for AI level role overrides

Creating  role overrides that only apply to one install allows to better handle edge cases where one school gives different meanings to the standar roles.

This PR:

- Provides a way to store this overrides (with the new table)
- Provide a method to view/add/delete/update overrides in the admin pages
- Keeps an audit trail of changes made to these.



### Testing

- Run the migration to get the new table:

 
```tox -e dev --run-command 'alembic upgrade head'```


Sanity check a few existing assignments:

- As `eng+canvasteacher@hypothes.is` launch: https://hypothesis.instructure.com/courses/125/assignments/873

You'll ge the teacher UI (ie you see the multiple section selection)


- As `eng+canvasstudent@hypothes.is` (in a incognito tab?)


launch https://hypothesis.instructure.com/courses/125/assignments/873

You'll get the student UI, no section selector.



- Check existing roles overrides for the AI:


Go to http://localhost:8001/admin/instance/8/, pick the Role Overrides tab. It's empty



- Create a new role override

Click `Create New override` pick: `Instructor` value and assign it to `Course` and `Learner`



- Check the override took effect, as the instructor, refresh: https://hypothesis.instructure.com/courses/125/assignments/873
It will only show one section now.



- Modify the override:
 
Go back to http://localhost:8001/admin/instance/8/ -> Role Overrides and select the existing override. 

Change the scope to "Institution".


-  Check the override took effect, as the instructor, refresh: https://hypothesis.instructure.com/courses/125/assignments/873
It will only show the section selector again. The institution scope means "somewhere else, but not in the current course".


- Finally, remove the override:

Go back to http://localhost:8001/admin/instance/8/ -> Role Overrides and select the existing override. 

Click delete at the bottom.


-  Check the override took effect
 
as the instructor, refresh: https://hypothesis.instructure.com/courses/125/assignments/873
It will only show the section selector again. This was the original behaviour with not overrides.


- Check the audit trail of these changes:

```docker compose exec postgres psql -U postgres -c "select extra->'action', extra->'userid', extra->'changes' from event_data where extra->>'model' = 'LTIRoleOverride'"```


```
 ?column? |          ?column?           |                                                     ?column?                                                     
----------+-----------------------------+------------------------------------------------------------------------------------------------------------------
"insert" | "marcos.prieto@hypothes.is" | {"type": [null, "learner"], "scope": [null, "course"], "lti_role": [null, 2], "application_instance": [null, 8]}
 "update" | "marcos.prieto@hypothes.is" | {"scope": ["course", "institution"]}
 "delete" | "marcos.prieto@hypothes.is" | {}
(3 rows)
```
